File: project/CCNet.VBDemos.Plugin/SourceControl/IndexFileSourceControl.vb
Imports System
Imports System.IO
Imports Exortech.NetReflector
Imports ThoughtWorks.CruiseControl.Core

<ReflectorType("indexFileSource")> Public Class IndexFileSourceControl
    Implements ISourceControl

    Private myFileName As String
    <ReflectorProperty("file")> _
    Public Property FileName() As String
        Get
            Return myFileName
        End Get
        Set(ByVal value As String)
            myFileName = value
        End Set
    End Property

    Public Function GetModifications(ByVal from As IIntegrationResult, _
                                     ByVal [to] As IIntegrationResult) As Modification() _
                                     Implements ISourceControl.GetModifications
        Dim path = [to].BaseFromWorkingDirectory(FileName)
        Dim reader = New StreamReader(path)
        Try
            Dim parser = New IndexFileHistoryParser()
            Return parser.Parse(reader, from.StartTime, [to].StartTime)
        Finally
            reader.Dispose()
        End Try
    End Function

    Public Sub GetSource(ByVal result As IIntegrationResult) _
    Implements ISourceControl.GetSource

    End Sub

    Public Sub Initialize(ByVal project As IProject) _
    Implements ISourceControl.Initialize

    End Sub

    Public Sub LabelSourceControl(ByVal result As IIntegrationResult) _
    Implements ISourceControl.LabelSourceControl

    End Sub

    Public Sub Purge(ByVal project As IProject) _
    Implements ISourceControl.Purge

    End Sub
End Class

File: project/CCNet.CSharpDemos.Plugin/SourceControl/IndexFileSourceControl.cs
using System;
using System.IO;
using Exortech.NetReflector;
using ThoughtWorks.CruiseControl.Core;

namespace CCNet.CSharpDemos.Plugin.SourceControl
{
    [ReflectorType("indexFileSource")]
    public class IndexFileSourceControl
        : ISourceControl
    {
        [ReflectorProperty("file")]
        public string FileName { get; set; }

        public Modification[] GetModifications(IIntegrationResult from, IIntegrationResult to)
        {
            var path = to.BaseFromWorkingDirectory(this.FileName);
            using (var reader = new StreamReader(path))
            {
                var parser = new IndexFileHistoryParser();
                return parser.Parse(reader, from.StartTime, to.StartTime);
            }
        }

        public void LabelSourceControl(IIntegrationResult result)
        {
            var fileName = Path.Combine(
                Path.GetDirectoryName(this.FileName),
                DateTime.Now.ToString("yyyyMMddHHmmss") + ".label");
            File.WriteAllText(fileName, result.Label);
        }

        public void GetSource(IIntegrationResult result)
        {
            foreach (var modification in result.Modifications)
            {
                var source = Path.Combine(
                    modification.FolderName,
                    modification.FileName);
                var destination = result.BaseFromWorkingDirectory(
                    modification.FileName);
                if (File.Exists(source))
                {
                    File.Copy(source, destination, true);
                }
                else
                {
                    File.Delete(destination);
                }
            }
        }

        public void Initialize(IProject project)
        {
            // Not needed
        }

        public void Purge(IProject project)
        {
            // Not needed
        }
    }
}
